vdr/ambassador.go

```diff
@@ -37,20 +49,245 @@ type Ambassador interface {
+	didStore      types.Store
+	keyResolver   crypto2.KeyResolver
 }
 
 // NewAmbassador creates a new Ambassador,
-func NewAmbassador(networkClient network.Network) Ambassador {
-	instance := &ambassador{
+func NewAmbassador(networkClient network.Network, didStore types.Store, keyResolver crypto2.KeyResolver) Ambassador {
+	return &ambassador{
 		networkClient: networkClient,
+		didStore:      didStore,
+		keyResolver:   keyResolver,
 	}
-	return instance
 }
 
+// NewDocumentVersion contains the version number that a new Network Documents have.
+const NewDocumentVersion = 0
+
 // Start instructs the ambassador to start receiving DID Documents from the network.
 func (n *ambassador) Start() {
-	n.networkClient.Subscribe(documentType, func(document dag.Document, payload []byte) error {
-		logging.Log().Warn("Not implemented: processing DID documents received from Nuts Network.")
-		return nil
-	})
+	n.networkClient.Subscribe(DIDDocumentType, n.callback)
+}
+
+// thumbprintAlg is used for creating public key thumbprints
+var thumbprintAlg = crypto.SHA256
+
+// callback gets called when new DIDDocuments are received by the network. All checks on the signature are already performed.
+// This method will check the integrity of the DID document related to the public key used to sign the network document.
+// The rules are based on the Nuts RFC006
+// payload should be a json encoded did.document
+func (n *ambassador) callback(document dag.SubscriberDocument, payload []byte) error {
+	logging.Log().Info("Processing DID documents received from Nuts Network.", DIDDocumentType, document.Ref())
+	if err := checkSubscriberDocumentIntegrity(document); err != nil {
+		return fmt.Errorf("callback could not process new DID Document: %w", err)
+	}
+
+	// Unmarshal the next/new proposed version of the DID Document
+	var nextDIDDocument did.Document
+	if err := json.Unmarshal(payload, &nextDIDDocument); err != nil {
+		return fmt.Errorf("unable to unmarshall did document from network payload: %w", err)
+	}
+
+	if isUpdate(document) {
+		return n.updateDIDDocument(document, nextDIDDocument)
+	}
+	return n.createDIDDocument(document, nextDIDDocument)
+}
+
+func (n *ambassador) createDIDDocument(document dag.SubscriberDocument, nextDIDDocument did.Document) error {
+	// First, check if the network document was signed by the same key as is embedded in the DID Document`s authenticationMethod:
+
+	// Take signing key from network document header
+	headerKey := document.SigningKey()
+	// Create key thumbprint
+	headerKeyThumbprint, err := headerKey.Thumbprint(thumbprintAlg)
+	if err != nil {
+		return fmt.Errorf("unable to generate network document signing key thumbprint: %w", err)
+	}
+
+	// Check if headerKey is one of the keys embedded in the authenticationMethod
+	didDocumentAuthKeys := nextDIDDocument.Authentication
+	if documentKey, err := n.findKeyByThumbprint(headerKeyThumbprint, didDocumentAuthKeys); documentKey == nil || err != nil {
```

Comment:
Now all errors are mapped to the "key not found" error as well. Imo it should be written something along the lines of;

```
if documentKey, err := n.findKeyByThumbprint(headerKeyThumbprint, didDocumentAuthKeys); err != nil {
    return err
} else if documentKey == nil {
	return fmt.Errorf("key used to sign Network document must be be part of DID Document authentication")
}
```